demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -432,46 +794,46 @@ static void _responseCompleteCallback( void * pPrivData,
 
-    /* If there is an error in the response processing. */
+    /* If there is was an error anywhere in the request response processing, including a cancellation, then that will
+     * end the demo. */
     if( rc != IOT_HTTPS_OK )
     {
         IotLogError( "There was a problem with the current response %d. Error code: %d. ", respHandle, rc );
         IotSemaphore_Post( &( _fileFinishedSem ) );
     }
-    else if( pDownloadData->currDownloaded != pDownloadData->numReqBytes )
+    else if( pRequestData->currDownloaded != pRequestData->numReqBytes )
     {
-        /* If in this response the total amount read does not equal the number of bytes we requested, then something went
-         * wrong. */
-        IotLogError( "There was a problem downloading the file. We downloaded %d. but wanted %d.",
-                     pDownloadData->currDownloaded,
-                     pDownloadData->numReqBytes );
+        /* If in this response the total amount read does not equal the number of bytes we requested, then something
+         * went wrong. */
+        IotLogError( "There was a problem downloading the range of the file. We downloaded %d. but wanted %d.",
+                     pRequestData->currDownloaded,
+                     pRequestData->numReqBytes );
         IotSemaphore_Post( &( _fileFinishedSem ) );
     }
     else
     {
-        /* The main application is waiting for the response to finish. We signal it is finished when the bytes
-         * downloaded so far is equal to the file size we expect. If there were problems in any of the async requests
-         * scheduled, then the demo will timeout anyways. */
-        _bytesFileDownloadedSoFar += pDownloadData->currDownloaded;
-        IotLogInfo( "Downloaded: %d/%d", _bytesFileDownloadedSoFar, _fileSize );
+        /* The main application is waiting for the response to finish. The demo application task signals that the file
+         * is finished when the number of ranges downloaded equals the total number of ranges. */
+        IotLogInfo( "Downloaded: %d/%d range blocks", _fileDownloadInfo.totalRanges - _fileDownloadInfo.rangesRemaining, _fileDownloadInfo.totalRanges );
 
-        if( _bytesFileDownloadedSoFar >= _fileSize )
+        if( _fileDownloadInfo.rangesRemaining <= 0 )
         {
-            IotLogDebug( "File fully downloaded. Bytes downloaded: %d", _bytesFileDownloadedSoFar );
+            IotLogDebug( "File fully downloaded. Range blocks downloaded: %d", _fileDownloadInfo.totalRanges - _fileDownloadInfo.rangesRemaining );
             IotSemaphore_Post( &( _fileFinishedSem ) );
         }
     }
+
+    /* Free up this request from the request pool. This is done last to ensure that the requestData is not
+     * overwritten by a new request. */
+    _freeRequestIndex( pRequestData->reqNum );
+    _fileDownloadInfo.scheduledBitmap[ BYTE_OFFSET( pRequestData->currRange ) ] |= BITMASK( pRequestData->currRange );
```

Comment:
Oops I meant:
```
_fileDownloadInfo.scheduledBitmap[ BYTE_OFFSET( pRequestData->currRange ) ] &= ~( BITMASK( pRequestData->currRange ) );
```

The original is OK because only a range that is both unfinished and unscheduled is allocated to a request.